javalanglib/src/main/scala/java/lang/Character.scala

```diff
@@ -519,6 +519,375 @@ object Character {
   @inline def compare(x: scala.Char, y: scala.Char): Int =
     x - y
 
+  // UnicodeBlock
+
+  class Subset protected (name: String) {
+    override final def equals(that: Any): scala.Boolean = super.equals(that)
+    override final def hashCode(): scala.Int = super.hashCode
+    override final def toString(): String = name
+  }
+
+  final class UnicodeBlock private (name: String, private val start: Int, private val end: Int) extends Subset(name)
+
+  object UnicodeBlock {
+    // Initial size from script below
+    private[this] val allBlocks: ArrayList[UnicodeBlock] = new ArrayList[UnicodeBlock](221)
+    private[this] val blocksByNormalizedName = new HashMap[String, UnicodeBlock]()
+
+    private[this] def addNameAliases(properName: String, block: UnicodeBlock,
+        includeJVM: scala.Boolean = true): Unit = {
+      // Add normalized aliases
+      val lower = properName.toLowerCase
+      //   lowercase + spaces
+      blocksByNormalizedName.put(lower, block)
+      //   lowercase + no spaces
+      blocksByNormalizedName.put(lower.replace(" ", ""), block)
+      //   tokens as _
+      if (includeJVM) {
+        blocksByNormalizedName.put(
+            lower.replace(' ', '_')
+              .replace('-', '_'), block)
+      }
+    }
+
+    private[this] def addUnicodeBlock(properName: String, start: Int, end: Int): UnicodeBlock =  {
+      val jvmName = properName.toUpperCase
+        .replace(' ', '_')
+        .replace('-', '_')
+
+      val block = new UnicodeBlock(jvmName, start, end)
+      allBlocks.add(block)
+      addNameAliases(properName, block)
+      block
+    }
+
+    // Special JVM Constant, don't add to allBlocks
+    val SURROGATES_AREA = new UnicodeBlock("SURROGATES_AREA", 0x0, 0x0)
+    blocksByNormalizedName.put("surrogates_area", SURROGATES_AREA)
+
+/*
+    // Get the "proper name" for JVM block name
+    val blockNameMap: Map[String, String] = {
+      val blocksSourceURL = new java.net.URL("http://unicode.org/Public/UCD/latest/ucd/Blocks.txt")
+      val source = scala.io.Source.fromURL(blocksSourceURL, "UTF-8")
+      source
+        .getLines()
+        .filterNot {
+          _.startsWith("#")
+        }
+        .flatMap { line =>
+          line.split(';') match {
+            case Array(_, name) =>
+              val trimmed = name.trim
+              val jvmName = trimmed.replaceAll(raw"[\s\-]", "_").toUpperCase
+              Some(jvmName -> trimmed)
+            case _ => None
+          }
+        }.toMap ++ Seq(
+          "GREEK" -> "Greek",
+          "CYRILLIC_SUPPLEMENTARY" -> "Cyrillic Supplementary",
+          "COMBINING_MARKS_FOR_SYMBOLS" -> "Combining Marks For Symbols"
+        )
```

Comment:
This is gonna sound annoying, since we have been back and forth about this for a while: It seems that these are the cases where the JVM name cannot be inferred from the real name of the block. So for me this means:

- Trying to find blocks based on the name is wrong, we should use its boundaries instead.
- Having a separate JVM name argument to `addUnicodeBlock` is OK, since there are genuine cases where these are different.

@sjrd WDYT?

Reply:
If I understand correctly from the JavaDoc, `Greek` and `Combining Marks For Symbols` have a different JVM name (+ alias with spaces) because those blocks *used to* be called like that. They were renamed in Unicode at some point (prior to 4.1.0 apparently; I could not find Blocks.txt for earlier versions than 4.1.0) so the new canonical names are usable as aliases, but the JVM name was kept.

For `Cyrillic Supplementary` it's unclear what the reason is. The canonical name in Unicode is `Cyrillic Supplement`, which is consistent with other `* Supplement` blocks. Regardless of the reason, it is also true that it can be found with the canonical name as well as with the JVM name and the JVM name with spaces.

I believe those are historical exceptions, and that more of those will not happen. The overwhelming rule still seems to be that the JVM name is directly derived from the canonical name. For those three exceptions, I would recommend having an overload of `addUnicodeBlock` that takes an additional `historicalName` and which registers the names differently (all lower-cased, obviously):

* the `properName`
* the `historicalName` (`"Greek"`, `"Cyrillic Supplementary"` or `"Combining Marks for Symbols"`)
* the `historicalName.replace(' ', '_')` (which is the name of the JVM constant)